native/dlaqr5.go

```diff
@@ -0,0 +1,609 @@
+//     update far-from-diagonal matrix entries.
+//  2: Dlaqr5 will accumulate reflections, use matrix-matrix multiply to update
+//     far-from-diagonal matrix entries, and take advantage of 2×2 block
+//     structure during matrix multiplies.
+// For other values of kacc2 Dlaqr5 will panic.
+//
+// n is the order of the Hessenberg matrix H.
+//
+// ktop and kbot are indices of the first and last row and column of an isolated
+// diagonal block upon which the QR sweep will be applied. It must hold that
+//  ktop == 0,   or 0 < ktop <= n-1 and H[ktop, ktop-1] == 0, and
+//  kbot == n-1, or 0 <= kbot < n-1 and H[kbot+1, kbot] == 0,
+// otherwise Dlaqr5 will panic.
+//
+// nshfts is the number of simultaneous shifts. It must be positive and even,
+// otherwise Dlaqr5 will panic.
+//
+// sr and si contain the real and imaginary parts, respectively, of the shifts
+// of origin that define the multi-shift QR sweep. On return both slices may be
+// reordered by Dlaqr5. Their length must be equal to nshfts, otherwise Dlaqr5
+// will panic.
+//
+// h and ldh represent the Hessenberg matrix H of size n×n. On return
+// multi-shift QR sweep with shifts sr+i*si has been applied to the isolated
+// diagonal block in rows and columns ktop through kbot, inclusive.
+//
+// iloz and ihiz specify the rows of Z to which transformations will be applied
+// if wantz is true. It must hold that 0 <= iloz <= ihiz < n, otherwise Dlaqr5
+// will panic.
+//
+// z and ldz represent the matrix Z of size n×n. If wantz is true, the QR sweep
+// orthogonal similarity transformation is accumulated into
+// z[iloz:ihiz,ilo:ihi] from the right, otherwise z not referenced.
+//
+// v and ldv represent an auxiliary matrix of size (nshfts/2)×3.
+//
+// u and ldv represent an auxiliary matrix of size (3*nshfts-3)×(3*nshfts-3).
+//
+// wh and ldwh represent an auxiliary matrix of size (3*nshfts-3)×nh.
+//
+// wv and ldwv represent an auxiliary matrix of size nv×(3*nshfts-3).
+func (impl Implementation) Dlaqr5(wantt, wantz bool, kacc22 int, n, ktop, kbot, nshfts int, sr, si []float64, h []float64, ldh int, iloz, ihiz int, z []float64, ldz int, v []float64, ldv int, u []float64, ldu int, nh int, wh []float64, ldwh int, nv int, wv []float64, ldwv int) {
+	checkMatrix(n, n, h, ldh)
+	if !(0 <= ktop && ktop < n) {
+		panic("lapack: invalid value of ktop")
+	}
+	if ktop > 0 && h[ktop*ldh+ktop-1] != 0 {
+		panic("lapack: diagonal block is not isolated")
+	}
+	if !(0 <= kbot && kbot < n) {
+		panic("lapack: invalid value of kbot")
+	}
+	if kbot < n-1 && h[(kbot+1)*ldh+kbot] != 0 {
+		panic("lapack: diagonal block is not isolated")
+	}
+	if nshfts < 0 || nshfts%2 != 0 {
+		panic("lapack: invalid number of shifts")
+	}
+	if len(sr) != nshfts || len(si) != nshfts {
+		panic(badSlice) // TODO(vladimir-ch) Another message?
+	}
+	if wantz {
+		if !(0 <= iloz && iloz <= ihiz && ihiz < n) {
+			panic("lapack: invalid value of iloz or ihiz")
```

Comment:
We can split the condition above and differentiate the two. This would clarify the message and make the condition nicer.

```
if ihiz >= n {
    panic("lapack: invalid value of ihiz")
}
if iloz <= 0 || ihiz < iloz {
    panic("lapack: invalid value of iloz")
}
```